components/net/websocket_loader.rs

```diff
@@ -105,80 +116,555 @@ pub fn init(connect: WebSocketCommunicate, connect_data: WebSocketConnectData, c
+            return Err(NetworkError::Internal(
+                "Response's Sec-WebSocket-Protocol header is missing, malformed or empty.".into()));
+        }
+    }
 
-    // Step 12, 14.
-    try!(response.validate());
+    // Step 14.2.
+    let upgrade_header = response.headers.get::<Upgrade>().ok_or_else(|| {
+        NetworkError::Internal("Response should have an Upgrade header.".into())
+    })?;
+    if upgrade_header.len() != 1 {
+        return Err(NetworkError::Internal("Response's Upgrade header should have only one value.".into()));
+    }
+    if upgrade_header[0].name != ProtocolName::WebSocket {
+        return Err(NetworkError::Internal("Response's Upgrade header value should be \"websocket\".".into()));
+    }
 
-    // Step 13 and transitive subset of step 14.
-    // See step 6 of http://tools.ietf.org/html/rfc6455#section-4.1.
-    let protocol_in_use = response.protocol().and_then(|header| {
-        // https://github.com/whatwg/fetch/issues/515
-        header.first().cloned()
+    // Step 14.3.
+    let connection_header = response.headers.get::<Connection>().ok_or_else(|| {
+        NetworkError::Internal("Response should have a Connection header.".into())
+    })?;
+    let connection_includes_upgrade = connection_header.iter().any(|option| {
+        match *option {
+            ConnectionOption::ConnectionHeader(ref option) => *option == "upgrade",
+            _ => false,
+        }
     });
-    if let Some(ref protocol_name) = protocol_in_use {
-        if !protocols.is_empty() && !protocols.iter().any(|p| (&**p).eq_ignore_ascii_case(protocol_name)) {
-            return Err(WebSocketError::ProtocolError("Protocol in Use not in client-supplied protocol list"));
-        };
+    if !connection_includes_upgrade {
+        return Err(NetworkError::Internal("Response's Connection header value should include \"upgrade\".".into()));
+    }
+
+    // Step 14.4.
+    let accept_header = response.headers.get::<WebSocketAccept>().ok_or_else(|| {
+        NetworkError::Internal("Response should have a Sec-Websocket-Accept header.".into())
+    })?;
+    if *accept_header != WebSocketAccept::new(&key_value) {
+        return Err(NetworkError::Internal(
+            "Response's Sec-WebSocket-Accept header value did not match the sent key.".into()));
+    }
+
+    // Step 14.5.
+    // TODO: handle permessage-deflate extension.
+    // We don't support any extension, so we fail at the mere presence of
+    // a Sec-WebSocket-Extensions header.
+    if response.headers.get_raw("Sec-WebSocket-Extensions").is_some() {
+        return Err(NetworkError::Internal(
+            "Response's Sec-WebSocket-Extensions header value included unsupported extensions.".into()));
+    }
+
+    // Step 14.6.
+    let protocol_in_use = if let Some(response_protocols) = response.headers.get::<WebSocketProtocol>() {
+        for p in &**response_protocols {
+            if !protocols.contains(p) {
```

Comment:
The previous check used eq_ignore_ascii_case. Is this the cause of the new test failure?